% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage{bookmark}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={scDEcrypter: Detecting latent viral states in scRNA-seq},
  pdfauthor={Luer Zhong},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{scDEcrypter: Detecting latent viral states in scRNA-seq}
\author{Luer Zhong}
\date{2025-11-20}

\begin{document}
\maketitle

\section{Introduction}\label{introduction}

Detecting infected cells in viral scRNA-seq data is challenging because
viral transcripts are often sparse or undetected, leaving most infection
labels unknown and limiting downstream differential expression analyses.
scDEcrypter is a statistical framework we developed to address this
problem by modeling infection status, and optionally other partitioning
variables such as cell type. The method uses a penalized multiway
mixture model, anchored by a small set of confidently labeled cells, to
recover latent infection states with high accuracy even when viral reads
are extremely limited. To avoid double-dipping, scDEcrypter employs a
data-splitting strategy, using one subset of cells for parameter
estimation and another independent subset for differential expression
testing. A fast approximate likelihood ratio test is then used to
identify infection-associated genes within each cell type. scDEcrypter
can improve power, robustness, and biological interpretability. If you
use scDEcrypter in published research, please cite:

\section{Run scDEcrypter}\label{run-scdecrypter}

\subsection{Install the package}\label{install-the-package}

Before analysis can be performed, scDEcrypter must be installed. The
easiest way to install scDEcrypter is through github:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(devtools)}
\NormalTok{devtools}\SpecialCharTok{::}\FunctionTok{install\_github}\NormalTok{(}\StringTok{"https://github.com/LZHONG25/scDEcrypter"}\NormalTok{)}
\FunctionTok{library}\NormalTok{(scDEcrypter)}
\end{Highlighting}
\end{Shaded}

\subsection{Required inputs}\label{required-inputs}

Load the Seurat object. We analyzed the COVID-19 nasal wash dataset
described in\\
Gao, Kevin M., \emph{et al.} (2021). \emph{Human nasal wash RNA-Seq
reveals distinct cell-specific innate immune responses in influenza
versus SARS-CoV-2}. \textbf{JCI Insight}, 6(22): e152288.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{seed }\OtherTok{\textless{}{-}} \FunctionTok{sample}\NormalTok{(}\DecValTok{63321232}\NormalTok{,}\DecValTok{1}\NormalTok{)}
\FunctionTok{set.seed}\NormalTok{(seed)}
\FunctionTok{load}\NormalTok{(}\StringTok{".../covid\_full\_raw\_seu.Rdata"}\NormalTok{)}
\CommentTok{\# the seurat object covid\_full\_raw\_seu.Rdata is saved as seurat\_obj}
\end{Highlighting}
\end{Shaded}

\subsubsection{Define the partitioning
variable}\label{define-the-partitioning-variable}

scDEcrypter requires one additional cell-level label, which we refer to
as the partitioning variable. This variable defines groups of cells that
are expected to differ in their baseline expression patterns, and it
helps the model separate biological heterogeneity unrelated to infection
status. Importantly, the partitioning variable is flexible and does not
need to be cell type. It can represent any categorical cell-level
attribute that the user believes is relevant.

In the example shown here, we use broad cell-type categories as the
partitioning variable. Fine-grained annotations from the Seurat object
are collapsed into six coarse groups (T cells, macrophages, B cells,
epithelial cells, dendritic cells, and neutrophils). Users may adapt or
replace this mapping entirely depending on the structure of their own
dataset. Make sure that the partitioning variable is numerically coded.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{seurat\_obj}\SpecialCharTok{$}\NormalTok{CellType\_Label }\OtherTok{\textless{}{-}} \FunctionTok{ifelse}\NormalTok{(seurat\_obj}\SpecialCharTok{$}\NormalTok{CellType\_Fine }\SpecialCharTok{\%in\%}
    \FunctionTok{c}\NormalTok{(}\StringTok{"CD4\_T"}\NormalTok{, }\StringTok{"CD8\_T"}\NormalTok{, }\StringTok{"prolifT"}\NormalTok{), }\DecValTok{1}\NormalTok{, }\FunctionTok{ifelse}\NormalTok{(seurat\_obj}\SpecialCharTok{$}\NormalTok{CellType\_Fine }\SpecialCharTok{\%in\%}
    \FunctionTok{c}\NormalTok{(}\StringTok{"alveol{-}mac"}\NormalTok{, }\StringTok{"IFNexp{-}mac"}\NormalTok{, }\StringTok{"M1{-}mac"}\NormalTok{,}
        \StringTok{"M1{-}mac{-}exp"}\NormalTok{, }\StringTok{"M2{-}mac"}\NormalTok{), }\DecValTok{2}\NormalTok{, }\FunctionTok{ifelse}\NormalTok{(seurat\_obj}\SpecialCharTok{$}\NormalTok{CellType\_Fine }\SpecialCharTok{\%in\%}
    \FunctionTok{c}\NormalTok{(}\StringTok{"memoryB1"}\NormalTok{, }\StringTok{"naiveB"}\NormalTok{, }\StringTok{"transB"}\NormalTok{, }\StringTok{"plasma"}\NormalTok{,}
        \StringTok{"ABC"}\NormalTok{), }\DecValTok{3}\NormalTok{, }\FunctionTok{ifelse}\NormalTok{(seurat\_obj}\SpecialCharTok{$}\NormalTok{CellType\_Fine }\SpecialCharTok{\%in\%}
    \FunctionTok{c}\NormalTok{(}\StringTok{"basal"}\NormalTok{, }\StringTok{"ciliated"}\NormalTok{, }\StringTok{"goblet+club"}\NormalTok{,}
        \StringTok{"hillock"}\NormalTok{), }\DecValTok{4}\NormalTok{, }\FunctionTok{ifelse}\NormalTok{(seurat\_obj}\SpecialCharTok{$}\NormalTok{CellType\_Fine }\SpecialCharTok{\%in\%}
    \FunctionTok{c}\NormalTok{(}\StringTok{"pDC"}\NormalTok{), }\DecValTok{5}\NormalTok{, }\FunctionTok{ifelse}\NormalTok{(seurat\_obj}\SpecialCharTok{$}\NormalTok{CellType\_Fine }\SpecialCharTok{\%in\%}
    \FunctionTok{c}\NormalTok{(}\StringTok{"G5a\_naive"}\NormalTok{, }\StringTok{"G5b"}\NormalTok{, }\StringTok{"G5c\_aged"}\NormalTok{, }\StringTok{"G5c\_naive"}\NormalTok{),}
    \DecValTok{6}\NormalTok{, }\ConstantTok{NA}\NormalTok{))))))}
\end{Highlighting}
\end{Shaded}

\subsubsection{Define the viral status
variable}\label{define-the-viral-status-variable}

scDEcrypter requires a cell-level viral status variable (V.obs) that
indicates whether each cell is infected or uninfected. In real scRNA-seq
viral infection datasets, this information is rarely fully observed:

(1). a small subset of cells have strong viral read evidence and can be
confidently labeled as infected,

(2). some cells (e.g., from unexposed or healthy samples) can be
confidently labeled as uninfected,

(3). some cells fall somewhere in between, where the absence of viral
reads does not guarantee that the cell is truly uninfected (e.g., low
viral load or bystander cells).

scDEcrypter is explicitly designed to model this partially latent
structure, as described in the paper. These confidently labeled cells
serve as anchors for the mixture model, while the remaining cells are
treated as unknown (NA) and their infection status is inferred
probabilistically during model fitting. To provide this information to
scDEcrypter, users should:

(1). assign 1 to confidently infected cells,

(2). assign 2 to confidently uninfected cells,

(3). leave all other cells as NA to allow scDEcrypter to infer their
latent infection status.

A typical setup might look like:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{C.obs }\OtherTok{\textless{}{-}}\NormalTok{ seurat\_obj}\SpecialCharTok{$}\NormalTok{CellType\_Label}

\NormalTok{V.obs }\OtherTok{\textless{}{-}} \FunctionTok{rep}\NormalTok{(}\ConstantTok{NA}\NormalTok{, }\FunctionTok{ncol}\NormalTok{(seurat\_obj))}
\NormalTok{to\_label\_I }\OtherTok{\textless{}{-}} \FunctionTok{which}\NormalTok{(seurat\_obj}\SpecialCharTok{$}\NormalTok{Infection\_byViralCounts }\SpecialCharTok{==} \StringTok{"Infected"}\NormalTok{)}
\NormalTok{to\_label\_U }\OtherTok{\textless{}{-}} \FunctionTok{which}\NormalTok{(seurat\_obj}\SpecialCharTok{$}\NormalTok{Group }\SpecialCharTok{==} \StringTok{"Healthy"}\NormalTok{)}
\NormalTok{V.obs[to\_label\_I] }\OtherTok{\textless{}{-}} \DecValTok{1}
\NormalTok{V.obs[to\_label\_U] }\OtherTok{\textless{}{-}} \DecValTok{2}
\end{Highlighting}
\end{Shaded}

\subsubsection{Define the number of infection states and partitioning
variable
classes}\label{define-the-number-of-infection-states-and-partitioning-variable-classes}

scDEcrypter models gene expression using a multiway mixture model, where
each cell belongs to a latent combination of a viral status class (e.g.,
infected, uninfected, bystander), and a partitioning variable class
(e.g., cell type, biological group, or any other cell-level label).

To fit this model, users must specify:

V.star: the total number of viral status categories the model should
consider,

C.star: the total number of categories for the chosen partitioning
variable.

These values define the full set of latent and observed combinations
that scDEcrypter will learn.

In many viral scRNA-seq datasets, infection status is not simply
``infected vs uninfected.'' Cells may exhibit transcriptional profiles
consistent with an intermediate `bystander' state, where they are not
infected but respond to nearby infected cells. As described in the
paper, for the COVID-19 nasal wash dataset, we observed: Infected cells
(with viral reads), Uninfected cells (from healthy donors), and
Bystander cells (from infected individuals, but lacking viral reads).
Because distinguishing bystander cells was biologically meaningful, we
set the total number of viral status categories as 3. This allows
scDEcrypter to infer these three viral states separately and perform
differential expression comparisons between them.

The partitioning variable (e.g., cell type) can also be partially
latent. Users must specify the total number of categories the model
should expect for that variable. For this dataset, after grouping
fine-grained cell annotations into broader categories, we determined
there were 6 cell type categories.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{V.star }\OtherTok{\textless{}{-}} \DecValTok{3}
\NormalTok{C.star }\OtherTok{\textless{}{-}} \DecValTok{6}
\end{Highlighting}
\end{Shaded}

\subsection{Run the pre-processing
function}\label{run-the-pre-processing-function}

To accurately recover latent infection states and perform unbiased
differential expression testing, scDEcrypter separates the dataset into
two independent subsets:

A generation set --- used only for parameter estimation in the penalized
multi-way mixture model;

A test set --- used only for downstream inference and likelihood-ratio
testing.

This split ensures that the same data are not used for both estimating
mixture parameters and testing hypotheses, thereby avoiding ``double
dipping,'' a key principle emphasized in the scDEcrypter paper.

The function preprocess\_scDEcrypter() automates all required
pre-processing steps:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  Splits cells into generation/test sets (stratified by known infection
  and partitioning labels when possible);
\item
  Applies normalization + variance stabilization independently to each
  split;
\item
  Returns variance-stabilized matrices for both sets;
\item
  Returns the indexing and label information required for model
  training.
\end{enumerate}

Variance stabilization is an essential part of the scDEcrypter workflow:
the underlying mixture model assumes approximately Gaussian stabilized
expression values. Users can specify their preferred method via
vs\_method, In this vignette, we use ``shifted\_log''.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{prep }\OtherTok{\textless{}{-}} \FunctionTok{preprocess\_scDEcrypter}\NormalTok{(seurat\_obj, C.obs, V.obs, seed, }\AttributeTok{vs\_method =} \StringTok{"shifted\_log"}\NormalTok{)}

\NormalTok{Y\_generation\_stbl }\OtherTok{\textless{}{-}}\NormalTok{ prep}\SpecialCharTok{$}\NormalTok{Y\_generation}
\NormalTok{Y\_test\_stbl }\OtherTok{\textless{}{-}}\NormalTok{ prep}\SpecialCharTok{$}\NormalTok{Y\_test}
\NormalTok{splitted\_data }\OtherTok{\textless{}{-}}\NormalTok{ prep}\SpecialCharTok{$}\NormalTok{splitted}
\end{Highlighting}
\end{Shaded}

scDEcrypter uses a penalized multi-way mixture model, which is
computationally efficient and statistically stable when trained on a
compact set of informative genes. Therefore, HVG selection is performed
only on the generation set, not the full dataset. This prevents
information leakage from the test set and preserves the integrity of the
data-splitting framework described in the manuscript. For the COVID
example, we used a cell-type--aware HVG selection procedure - within
each cell type, among COVID-positive samples only, and selected the most
variable genes per cell type, then aggregated them. Users may replace
the above with any HVG selection method, as long as HVGs are selected
only from the generation set, and the resulting HVG set is passed into
the model-fitting step. This ensures that downstream inference remains
valid and independent from the model training process.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{vargroup }\OtherTok{\textless{}{-}} \FunctionTok{na.omit}\NormalTok{(seurat\_obj}\SpecialCharTok{$}\NormalTok{CellType\_Label[splitted\_data}\SpecialCharTok{$}\NormalTok{generation\_idx])}
\NormalTok{covgroup }\OtherTok{\textless{}{-}} \FunctionTok{names}\NormalTok{(}\FunctionTok{which}\NormalTok{(seurat\_obj}\SpecialCharTok{$}\NormalTok{Group[splitted\_data}\SpecialCharTok{$}\NormalTok{generation\_idx] }\SpecialCharTok{==} \StringTok{"COVID{-}19"}\NormalTok{))}

\NormalTok{top\_genesA }\OtherTok{\textless{}{-}} \FunctionTok{sapply}\NormalTok{(}\FunctionTok{unique}\NormalTok{(vargroup), }\ControlFlowTok{function}\NormalTok{(x)}
  \FunctionTok{apply}\NormalTok{(Y\_generation\_stbl[}\FunctionTok{intersect}\NormalTok{(covgroup, }\FunctionTok{names}\NormalTok{(}\FunctionTok{which}\NormalTok{(vargroup}\SpecialCharTok{==}\NormalTok{x))),], }\DecValTok{2}\NormalTok{, var))}

\NormalTok{topidx }\OtherTok{\textless{}{-}} \FunctionTok{apply}\NormalTok{(top\_genesA, }\DecValTok{2}\NormalTok{, }\ControlFlowTok{function}\NormalTok{(x) }\FunctionTok{order}\NormalTok{(x, }\AttributeTok{decreasing =}\NormalTok{ T)[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{500}\NormalTok{])}
\NormalTok{topidx\_g }\OtherTok{\textless{}{-}} \FunctionTok{apply}\NormalTok{(topidx, }\DecValTok{2}\NormalTok{, }\ControlFlowTok{function}\NormalTok{(x) }\FunctionTok{rownames}\NormalTok{(top\_genesA)[x])}
\NormalTok{top\_genes.HVG }\OtherTok{\textless{}{-}} \FunctionTok{unique}\NormalTok{(}\FunctionTok{as.vector}\NormalTok{(}\FunctionTok{apply}\NormalTok{(topidx\_g, }\DecValTok{2}\NormalTok{, }\ControlFlowTok{function}\NormalTok{(x) x[}\DecValTok{1}\SpecialCharTok{:}\DecValTok{50}\NormalTok{])))}
\end{Highlighting}
\end{Shaded}

\subsection{Select the best tuning
parameter}\label{select-the-best-tuning-parameter}

scDEcrypter employs a penalized mixture modeling framework, where the
penalty parameter lambda controls the amount of regularization applied
to the model's mean parameters. Choosing an appropriate lambda is
essential for balancing model flexibility with stability, and for
recovering biologically meaningful latent infection states. To
accomplish this, scDEcrypter provides a built-in cross-validation
procedure that evaluates a set of candidate lambda values using the
held-out generation set created earlier.

Users may customize three aspects of lambda selection:

A grid of candidate lambda values: this grid can span large orders of
magnitude because the optimal penalty varies across datasets.

The maximum number of EM iterations (max.iter): controls how long the
model is allowed to iterate for each lambda value.

A convergence tolerance (tol): determines when the EM algorithm has
stabilized.

All other required inputs, such as stabilized expression values, label
vectors, and data splits, were already constructed in earlier steps.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{lambda.vec }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\FloatTok{1e2}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{0}\NormalTok{, }\FloatTok{1e{-}1}\NormalTok{, }\FloatTok{1e{-}2}\NormalTok{, }\FloatTok{1e{-}3}\NormalTok{, }\FloatTok{1e{-}4}\NormalTok{, }\FloatTok{1e{-}5}\NormalTok{, }\FloatTok{1e{-}6}\NormalTok{, }\FloatTok{1e{-}7}\NormalTok{, }\FloatTok{1e{-}8}\NormalTok{)}
\NormalTok{cv\_lambda }\OtherTok{\textless{}{-}} \FunctionTok{cross\_validate\_lambda}\NormalTok{(Y\_generation\_stbl[,top\_genes.HVG],}
\NormalTok{                                   splitted\_data}\SpecialCharTok{$}\NormalTok{C.obs\_generation,}
\NormalTok{                                   splitted\_data}\SpecialCharTok{$}\NormalTok{V.obs\_generation, }
\NormalTok{                                   lambda.vec, }\AttributeTok{max.iter =} \DecValTok{200}\NormalTok{, }
                                   \AttributeTok{tol =} \FloatTok{1e{-}8}\NormalTok{, C.star, V.star, seed, }
                                   \AttributeTok{NCORES =} \DecValTok{5}\NormalTok{)}
\NormalTok{best\_lambda }\OtherTok{\textless{}{-}}\NormalTok{ cv\_lambda}\SpecialCharTok{$}\NormalTok{best\_lambda}
\end{Highlighting}
\end{Shaded}

\subsection{Fit the scDEcrypter multiway mixture
model}\label{fit-the-scdecrypter-multiway-mixture-model}

After selecting the optimal penalty parameter, we fit the full
scDEcrypter model using the generation set. Recall that the generation
set was set aside specifically for model training to avoid using the
same data for both parameter estimation and downstream inference. In
this step, scDEcrypter estimates all components of the penalized
multiway mixture model described in the manuscript. These fitted
parameters are essential for:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  Predicting latent infection states for cells whose viral status is
  unknown (NA in V.obs)
\item
  Performing differential expression analysis based on the inferred
  infection groups
\end{enumerate}

The results\_generation object stores:

(1). posterior probabilities for latent infection states,

(2). estimated mixture proportions for each (infection × partition)
combination,

(3). penalized mean and variance parameters,

(4). convergence diagnostics and log-likelihood trajectories.

These estimates form the foundation for the next steps: predicting
infection status for all cells and conducting cell-type--specific
differential expression testing.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{results\_generation }\OtherTok{\textless{}{-}} \FunctionTok{MultiwayMixture}\NormalTok{(}\AttributeTok{Y =}\NormalTok{ Y\_generation\_stbl[,top\_genes.HVG],}
                                      \AttributeTok{C.obs =}\NormalTok{ splitted\_data}\SpecialCharTok{$}\NormalTok{C.obs\_generation,}
                                      \AttributeTok{V.obs =}\NormalTok{ splitted\_data}\SpecialCharTok{$}\NormalTok{V.obs\_generation,}
                                      \AttributeTok{max.iter =} \DecValTok{200}\NormalTok{, }\AttributeTok{tol =} \FloatTok{1e{-}8}\NormalTok{, }
\NormalTok{                                      C.star, V.star, seed, }
                                      \AttributeTok{lambda.vec=}\FunctionTok{dim}\NormalTok{(splitted\_data}\SpecialCharTok{$}\NormalTok{Y\_generation)[}\DecValTok{1}\NormalTok{]}
                                      \SpecialCharTok{*}\NormalTok{best\_lambda)}
\end{Highlighting}
\end{Shaded}

\section{Infection status prediction}\label{infection-status-prediction}

\subsection{Choosing cutoffs for infection-state
prediction}\label{choosing-cutoffs-for-infection-state-prediction}

After estimating the multiway mixture model on the generation set,
scDEcrypter predicts infection states for all cells by using their
posterior probabilities across the latent viral-status classes. These
posterior probabilities summarize how strongly each cell is supported as
``infected,'' ``uninfected,'' or (in settings like the COVID dataset)
``bystander,'' based on both its gene expression and the structure of
the mixture model.

To convert these posterior probabilities into labels, scDEcrypter allows
users to specify cutoff parameters:

(1). Posterior probability threshold: this cutoff controls how much
certainty is required before a cell is assigned to a particular
infection class. If a cell has posterior probability ≥ cutoffs for one
class, it is labeled as belonging to that class. A higher threshold
(e.g., 0.9) leads to more conservative labeling---only the cells with
the strongest evidence will be assigned.

(2). Minimum proportion of confidently labeled cells needed: this
threshold specifies how many cells in a class must pass the posterior
certainty requirement before scDEcrypter considers that class
sufficiently well represented for downstream differential expression
analysis. For example: ff fewer than 75\% of cells assigned to a class
meet the posterior cutoff, the class may be flagged as underpowered or
unstable. This helps ensure that DE inference is performed only when
scDEcrypter's latent-state predictions are reliable.

Users may adjust these values depending on the dataset and the desired
balance between sensitivity and certainty in infection-state
assignments.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cutoffs }\OtherTok{\textless{}{-}}\NormalTok{ .}\DecValTok{9}
\NormalTok{cell\_thresh }\OtherTok{\textless{}{-}}\NormalTok{ .}\DecValTok{75}
\end{Highlighting}
\end{Shaded}

\subsection{Predicting cell type and infection
status}\label{predicting-cell-type-and-infection-status}

After estimating the mixture model parameters on the generation set,
scDEcrypter next predicts (1) cell-type labels and (2) infection-state
labels for all cells in the test set. These predicted labels are then
used for downstream differential expression analysis.

The following steps extract the fitted mixture model parameters, compute
posterior probabilities for all latent states in the test set, and
convert them into hard labels.

\subsubsection{Extracting model
parameters}\label{extracting-model-parameters}

These objects contain the estimated quantities from the fitted penalized
two-way mixture model:

(1). M: estimated mean expression for each gene, cell-type, and
infection-status combination,

(2). sigma2: estimated variances,

(3). probs: estimated mixing proportions,

(4). W.gen: posterior weights for cells in the generation set,

(5). W.test: posterior weights for cells in the test set.

These parameters summarize the inferred latent structure and are used to
predict labels for new cells.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{M }\OtherTok{\textless{}{-}}\NormalTok{ results\_generation}\SpecialCharTok{$}\NormalTok{M\_list[[}\DecValTok{1}\NormalTok{]]}
\NormalTok{sigma2 }\OtherTok{\textless{}{-}}\NormalTok{ results\_generation}\SpecialCharTok{$}\NormalTok{sigma2\_list[[}\DecValTok{1}\NormalTok{]]}
\NormalTok{probs }\OtherTok{\textless{}{-}}\NormalTok{ results\_generation}\SpecialCharTok{$}\NormalTok{probs\_list[[}\DecValTok{1}\NormalTok{]]}
\NormalTok{W.gen }\OtherTok{\textless{}{-}}\NormalTok{ results\_generation}\SpecialCharTok{$}\NormalTok{weights\_list[[}\DecValTok{1}\NormalTok{]]}
\NormalTok{W.test }\OtherTok{\textless{}{-}} \FunctionTok{E.step1}\NormalTok{(Y\_test\_stbl[, top\_genes.HVG], splitted\_data}\SpecialCharTok{$}\NormalTok{C.obs\_test,}
\NormalTok{    splitted\_data}\SpecialCharTok{$}\NormalTok{V.obs\_test, M, probs, sigma2)}
\end{Highlighting}
\end{Shaded}

\subsubsection{Predicting cell type
labels}\label{predicting-cell-type-labels}

Step-by-step explanation:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  Collapse the infection dimension: apply(W.test, c(1, 2), max) finds
  for each cell and each cell type, the maximum probability across all
  infection states. This gives a matrix: cells × celltypes, containing
  how strongly each cell is supported as each cell type regardless of
  infection.
\item
  Assign cell-type labels: max.col() selects the cell-type with the
  highest posterior probability.
\item
  Apply certainty threshold: cells with overall certainty below
  cell\_thresh (e.g., 0.75) are set to 0 (unassigned), ensuring
  downstream DE tests rely only on confidently predicted cell types.
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{W\_max\_byC }\OtherTok{\textless{}{-}} \FunctionTok{apply}\NormalTok{(W.test, }\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{2}\NormalTok{), max, }\AttributeTok{na.rm =} \ConstantTok{TRUE}\NormalTok{)}
\NormalTok{CellType\_pred }\OtherTok{\textless{}{-}} \FunctionTok{max.col}\NormalTok{(W\_max\_byC, }\AttributeTok{ties.method =} \StringTok{"first"}\NormalTok{)}
\NormalTok{CellType\_pred[}\FunctionTok{which}\NormalTok{(}\FunctionTok{apply}\NormalTok{(W\_max\_byC, }\DecValTok{1}\NormalTok{, max) }\SpecialCharTok{\textless{}}\NormalTok{ cell\_thresh)] }\OtherTok{\textless{}{-}} \DecValTok{0}
\end{Highlighting}
\end{Shaded}

\subsubsection{Predicting infection status
labels}\label{predicting-infection-status-labels}

Step-by-step explanation:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  Collapse the cell type dimension: apply(W.test, c(1, 3), max) computes
  the maximum posterior probability across cell types for each infection
  state. This produces a cells × infection\_states matrix summarizing
  the evidence for each cell's infection class.
\item
  Initialize all cells as ``Unknown'': this allows us to update
  prediction labels only when the posterior probability exceeds the
  cutoff.
\item
  Assign infection labels by thresholding: for each infection state v,
  ff a cell has posterior probability ≥ cutoffs (e.g., 0.9), then
  scDEcrypter assigns the corresponding infection label.
\end{enumerate}

This logic reflects the biological uncertainty in viral studies: even if
only a subset of cell-type--specific probabilities are high, the cell
can confidently belong to that infection state.

Important notes: the ordering of state\_names must correspond to the
user's numerical encoding (e.g., 1 = Infected, 2 = Uninfected, 3 =
Bystander).

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{W.byV }\OtherTok{\textless{}{-}} \FunctionTok{apply}\NormalTok{(W.test, }\FunctionTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\DecValTok{3}\NormalTok{), max)}
\NormalTok{Full\_Prediction }\OtherTok{\textless{}{-}} \FunctionTok{rep}\NormalTok{(}\StringTok{"Unknown"}\NormalTok{, }\FunctionTok{nrow}\NormalTok{(W.byV))}
\NormalTok{state\_names }\OtherTok{\textless{}{-}} \FunctionTok{c}\NormalTok{(}\StringTok{"Infected"}\NormalTok{, }\StringTok{"Uninfected"}\NormalTok{, }\StringTok{"Bystander"}\NormalTok{)}
\ControlFlowTok{for}\NormalTok{ (v }\ControlFlowTok{in} \FunctionTok{seq\_len}\NormalTok{(}\FunctionTok{ncol}\NormalTok{(W.byV))) \{}
\NormalTok{  Full\_Prediction[W.byV[, v] }\SpecialCharTok{\textgreater{}=}\NormalTok{ cutoffs] }\OtherTok{\textless{}{-}}\NormalTok{ state\_names[v]}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{temp1 }\OtherTok{\textless{}{-}} \FunctionTok{as.data.frame}\NormalTok{(Full\_Prediction)}
\NormalTok{temp1}\SpecialCharTok{$}\NormalTok{CellType\_pred }\OtherTok{\textless{}{-}}\NormalTok{ CellType\_pred}
\NormalTok{temp1}\SpecialCharTok{$}\NormalTok{cellID }\OtherTok{\textless{}{-}}\NormalTok{ seurat\_obj}\SpecialCharTok{$}\NormalTok{cellID[splitted\_data}\SpecialCharTok{$}\NormalTok{test\_idx]}
\NormalTok{temp1 }\OtherTok{\textless{}{-}} \FunctionTok{merge}\NormalTok{(temp1, seurat\_obj}\SpecialCharTok{@}\NormalTok{meta.data[splitted\_data}\SpecialCharTok{$}\NormalTok{test\_idx,], }\AttributeTok{by=}\StringTok{"cellID"}\NormalTok{, }\AttributeTok{all.x=}\NormalTok{T, }\AttributeTok{sort=}\NormalTok{F)}
\end{Highlighting}
\end{Shaded}


\end{document}
